### PR TITLE
gql-tag-preset: support data/fragment masking

### PR DESCRIPTION
**Pull request by n1ru4l**
Relay style fragment (type! not runtime) masking with the gql-tag-operations preset.

### CODE REVIEW

File: packages/plugins/typescript/gql-tag-operations/src/index.ts [hunk 1]
```diff
@@ -21,13 +21,34 @@ export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocume
...
+      : never
+    : never
+  : never;
+
+export const useFragment = <TType>(
+  _documentNode: DocumentNode<TType, any>,
+  fragmentType: FragmentType<DocumentNode<TType, any>>
+): TType => {
+  return fragmentType as any;
+};
```
Comment (n1ru4l) [inline packages/plugins/typescript/gql-tag-operations/src/index.ts:41]: If frameworks wanna implement true fragment type masking at runtime they just need to implement this function type signature.

File: packages/plugins/typescript/gql-tag-operations/src/index.ts [hunk 2]
```diff
@@ -21,13 +21,34 @@ export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocume
...
+export type FragmentType<TDocumentType extends DocumentNode<any, any>> = TDocumentType extends DocumentNode<
+  infer TType,
+  any
+>
+  ? TType extends { ' $fragmentName': infer TKey }
+    ? TKey extends string
+      ? { ' $fragmentRefs': { [key in TKey]: TType } }
+      : never
+    : never
+  : never;
```
Comment (n1ru4l) [inline packages/plugins/typescript/gql-tag-operations/src/index.ts:34]: This is the type used to add a fragment dependency on a component

File: .changeset/five-donuts-enjoy.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+'@graphql-codegen/gql-tag-operations-preset': minor
+---
+
+The plugin now generates an `gql.ts` file for the `gql` function, whose contents are re-exported from the `index.ts` file. In module augmentation mode the `index.ts` file is omitted and only a `gql.d.ts` file is generated.
+
+Support for fragment masking via the new `fragmentMasking` configuration option. [Check out the Fragment Masking Documentation](https://envelop.dev/plugins/gql-tag-operations-preset#fragment-masking).
```
Comment (n1ru4l) [inline .changeset/five-donuts-enjoy.md:7]: nit: rename this to dataMasking to be more similar to relay naming https://relay.dev/docs/principles-and-architecture/thinking-in-relay/#data-masking

Comment (n1ru4l) [reply .changeset/five-donuts-enjoy.md:7]: I changed my mind.

**Review by dotansimha — APPROVED** [review]

File: .changeset/five-donuts-enjoy.md (hunk 3 above)
Comment (charlypoly) [inline .changeset/five-donuts-enjoy.md:7]: @n1ru4l this link (https://www.envelop.dev/plugins/gql-tag-operations-preset#fragment-masking) is a 404 👀

Comment (n1ru4l) [reply .changeset/five-donuts-enjoy.md:7]: lol this should be graphql-code-generator

**Comment by Coobaha** [comment]
@n1ru4l this changes are breaking types/or some assumptions of https://www.npmjs.com/package/graphql-codegen-typescript-operation-types in 2.3.1 for inline fragments. We are not using masks, and using only `typescript-operations` with https://www.npmjs.com/package/graphql-codegen-typescript-operation-types

And it starts outputting `{' $fragmentRefs': T }`  instead of actual types

fyi @Stonepaw

**Comment by n1ru4l** [comment]
@Coobaha Can you provide a minimal reproduction of this issue on CodeSandbox or stackblitz?

**Comment by Coobaha** [comment]
@n1ru4l Here is minimal repro https://codesandbox.io/s/jovial-bassi-ev01gi?file=/types.ts

**Comment by n1ru4l** [comment]
@Coobaha Working on a fix over here: https://github.com/dotansimha/graphql-code-generator/pull/7555

**Comment by n1ru4l** [comment]
@Coobaha You can already use `@graphql-codegen/typescript-operations@2.3.2-alpha-e6dadab0d.0` https://codesandbox.io/s/agitated-haze-5wdo1d?file=/types.ts - fix will be release soon :)

**Comment by n1ru4l** [comment]
@Coobaha fix is released as `@graphql-codegen/typescript-operations@2.3.2`!

**Comment by Coobaha** [comment]
Thanks for such a quick fix! 🥳

**Pull request merged**